### PR TITLE
Add support for collapsing on multiple fields

### PR DESCRIPTION
**Pull request by tomglk**
The FieldCollapsingSearcher is currently only able to collapse on a single field.

This PR adds support to collapse on an arbitrary amount of fields.

**State of the PR:**
~~The PR shows a possible solution, which needs feedback for the implemented approach.
Test Cases are missing, Code Review is needed, but later.~~

~~**First goal is to clarify if we agree on the general idea.**~~

We agree on the general idea.
Code Review was done.
Test Cases are in progress.

I think that multi-field collapsing can be understood in multiple ways:
- collapse by the fields **one by one**, looking at their values & collapsesizes separately after each other
- collapse by the **combination** of the fields, meaning the value-combinations of the fields all at once

This PR implements the first understanding.

---
**Changes in FieldCollapsingSearcher:**
- Allow passing a comma-separated list of fields in collapsefield
- Use collapsesize to override the default for all fields
- Use collapsesize.fieldname to override the size for a specific field
- collapse the result multiple times by iterating through the collapsefields

This means that the collapsesize for field `foo` is determined like this:
- IF collapsesize.foo is set: use that
- ELSE IF collapsesize is set: use that
- ELSE: use default collapsesize

---
**Closed topics:**
- The order of fields is important (see testCollapsingWithMultipleFields), opinions towards that?
    - This is fine.
- The parameter is still called `collapsefield` (to keep backwards-compatibility), is that a deal-breaker?
    - It is not.
- Regarding the different understandings from the top: Do you agree on those? Which should we support, both?
    - We support the first understanding - use the fields sequentially.

**Thoughts about the field-order topic:**
If the order of the fields would not matter, how should we determine which hits are kept? Should we try to keep as many hits per collapsefield as possible (reaching as close as possible to collapsesize)?
What happens if there is no clear "winner", meaning we can only get n hits, no matter how we combine the collapsefields?

In short: I think that the collapse-logic can become very complex if we don't simply take the fields in the order they are configured.

You are welcome to challenge that. :)

---
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

**Review by vespa-headless — COMMENTED** [review]
Great! Collapsing the fields one by one (or, locally not globally) sounds good to me, and I agree it should then depend on ordering.

Reusing the parameter names like this is also good.

(I agree some more testing is needed but otherwise the code LGTM.)

**Review by bratseth — COMMENTED** [review]
Great! Collapsing the fields one by one (or, locally not globally) sounds good to me, and I agree it should then depend on ordering.

Reusing the parameter names like this is also good.

(I agree some more testing is needed but otherwise the code LGTM.)

File: container-search/src/main/java/com/yahoo/prelude/searcher/FieldCollapsingSearcher.java [hunk 1]
```diff
@@ -159,19 +181,35 @@ private void collapse(Result result, Map<String, Integer> knownCollapses,
                 continue;
             }
 
-            if (knownCollapses.containsKey(collapseId)) {
-                int numHitsThisField = knownCollapses.get(collapseId);
+            // prepending the fieldname is necessary to distinguish between values in the different collapsefields
+            String collapseKey = collapseField + "_" + collapseId;
```
Comment (bratseth) [inline container-search/src/main/java/com/yahoo/prelude/searcher/FieldCollapsingSearcher.java:185]: Underscore is legal in field names so I suggest using e.g @.

**Comment by tomglk** [comment]
Update for the documentation: https://github.com/vespa-engine/documentation/pull/3051

**Review by bratseth — APPROVED** [review]
👍  Ready to merge?

**Comment by tomglk** [comment]
yes
